maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java

```diff
@@ -0,0 +1,144 @@
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.maven.plugins.packager.config;
+
+import static org.apache.commons.lang.StringUtils.isEmpty;
+import static org.openmrs.module.initializer.validator.Validator.ARG_CIEL_FILE;
+import static org.openmrs.module.initializer.validator.Validator.ARG_CONFIG_DIR;
+import static org.openmrs.module.initializer.validator.Validator.ARG_UNSAFE;
+
+import java.io.File;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+
+import javax.persistence.Cache;
+import javax.persistence.EntityGraph;
+import javax.persistence.PersistenceException;
+import javax.persistence.PersistenceUnitUtil;
+import javax.persistence.Query;
+import javax.persistence.TransactionRequiredException;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.metamodel.Metamodel;
+
+import org.apache.commons.cli.ParseException;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.junit.runner.Result;
+import org.openmrs.module.initializer.validator.Validator;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.orm.jpa.EntityManagerFactoryUtils;
+import org.springframework.orm.jpa.JpaObjectRetrievalFailureException;
+import org.springframework.orm.jpa.JpaOptimisticLockingFailureException;
+import org.springframework.orm.jpa.JpaSystemException;
+
+/**
+ * The purpose of this Mojo is to validate configurations.
+ */
+@Mojo( name = "validate-configurations" )
+public class ValidateConfigurationsMojo extends AbstractPackagerConfigMojo {
+	
+	// Configuration Directory
+	@Parameter(property = "sourceDir", defaultValue = "configuration")
+	private File sourceDir;
+	
+	@Parameter(property = "cielFile")
+	private File cielFile;
+	
+	// Extra Validator CLI options
+	@Parameter(property = "extraValidatorArgs")
+	private String extraValidatorArgs;
+	
+	/*
+	 * To avoid NoClassDefFoundError on EntityManagerFactoryUtils and related classes.
+	 */
+	private void findClassDefinitions() {
+		EntityManagerFactoryUtils.class.toString();
+		PersistenceException.class.toString();
+		TransactionRequiredException.class.toString();
+		JpaObjectRetrievalFailureException.class.toString();
+		EmptyResultDataAccessException.class.toString();
+		JpaOptimisticLockingFailureException.class.toString();
+		JpaSystemException.class.toString();
+		Cache.class.toString();
+		CriteriaBuilder.class.toString();
+		Metamodel.class.toString();
+		PersistenceUnitUtil.class.toString();
+		Query.class.toString();
+		EntityGraph.class.toString();
+	}
+	
+	/**
+	 * @throws MojoExecutionException if the Maven build should be errored.
+	 */
+	public void execute() throws MojoExecutionException {
+		
+		findClassDefinitions(); // TODO: figure out why this is needed
```

Comment:
@mseaton the biggest puzzle. Failing to do this triggers errors when the Spring context is destroyed. Everything works fine otherwise but the logs get bloated with weird linkage errors (if I remember correctly.)

Reply:
@mks-d I can try to poke around, but nothing springs immediately to mind.  Would be nice to figure out and clean up but shouldn't need to hold up the merge and getting this into the project.  As soon as the validator jar is generally available in iniz, and this will build cleanly, it looks good to me.

I wasn't able to test this breaking.  If I comment out the findClassDefinitions(); line and run a mvn clean install, everything builds and all unit tests pass.  Is there a way to reproduce?

Reply:
Thanks @mseaton, much appreciated.

>Is there a way to reproduce?

This is a runtime-only issue, not a build issue. If you configure the validate goal for a config in its POM and that you comment out `findClassDefinitions()` you should see the problem coming up.

Reply:
OK, will see if I can try it at some point.  I'll leave it to you to merge this in when ready based on when the validator jar is ready and available for this to build with.